apps/confidential/confidential.go

```diff
@@ -0,0 +1,364 @@
+	if secret == "" {
+		return Credential{}, errors.New("secret can't be empty string")
+	}
+	return Credential{secret: secret}, nil
+}
+
+// NewCredFromCert creates a Credential from an x509.Certificate and a PKCS8 DER encoded private key.
+// CertFromPEM() can be used to get these values from a PEM file storing a PKCS8 private key.
+func NewCredFromCert(cert *x509.Certificate, key crypto.PrivateKey) Credential {
+	return Credential{cert: cert, key: key}
+}
+
+// Client is a representation of authentication client for confidential applications as defined in the
+// package doc. A new Client should be created PER SERVICE USER.
+// For more information, visit https://docs.microsoft.com/azure/active-directory/develop/msal-client-applications
+type Client struct {
+	client.Base
+
+	cred *accesstokens.Credential
+
+	// userID is some unique identifier for a user. It actually isn't used by us at all, it
+	// simply acts as another hint that a confidential.Client is for a single user.
+	userID string
+}
+
+// Options are optional settings for New(). These options are set using various functions
+// returning Option calls.
+type Options struct {
+	// Accessor controls cache persistence.
+	// By default there is no cache persistence. This can be set using the Accessor() option.
+	Accessor cache.ExportReplace
+
+	// The host of the Azure Active Directory authority.
+	// The default is https://login.microsoftonline.com/common. This can be changed using the
+	// Authority() option.
+	Authority string
+}
+
+func (o Options) validate() error {
+	u, err := url.Parse(o.Authority)
+	if err != nil {
+		return fmt.Errorf("the Authority(%s) does not parse as a valid URL", o.Authority)
+	}
+	if u.Scheme != "https" {
+		return fmt.Errorf("the Authority(%s) does not appear to use https", o.Authority)
+	}
+	return nil
+}
+
+// Option is an optional argument to New().
+type Option func(o *Options)
+
+// Authority allows you to provide a custom authority for use in the client.
+func Authority(authority string) Option {
+	return func(o *Options) {
+		o.Authority = authority
+	}
+}
+
+// Accessor provides a cache accessor that will read and write to some externally managed cache
+// that may or may not be shared with other applications.
+func Accessor(accessor cache.ExportReplace) Option {
+	return func(o *Options) {
+		o.Accessor = accessor
+	}
+}
+
+// tokener has a shared requests.Token object. I (jdoak) am not a fan. But at this point, that
+// object is internal/ and I don't want to pull it out. A confidential.Client is mean to be made
+// per user, so we don't want to be creating a bunch of requests.Token objects.
+var tokener = requests.NewToken()
+
+// New is the constructor for Client. userID is the unique identifier of the user this client
+// will store credentials for (a Client is per user). clientID is the Azure clientID and cred is
+// the type of credential to use.
+func New(userID, clientID string, cred Credential, options ...Option) (Client, error) {
```

Comment:
While I get why you did this, I can't help but feel that the naming is now less intuitive/descriptive.

```go
// wordy but obvious
client := msal.NewConfidentialClientApplication()

// succinct but less clear
client := confidential.New()
```

Don't get me wrong, a lot of the type names before this change we just too verbose.  I just feel like we've swung the pendulum too far in the other direction.

I don't have a suggestion at present, and I'm fine with making such changes post-merge of this PR.

Reply:
What is `userID`?  This param wasn't in the previous iteration.

Reply:
see: https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/9dbe73b396b302c2daa32f24847745f7b63bd94b/apps/design/design.md#confidential-applications-have-a-user-id

Reply:
Naming is probably the hardest thing to get right, especially in Go where we believe in small names and using package structure as part of names.  And its a philosophical argument that will be happening long after I'm dead.  .Net and Java seem to think that StringObjectThatRepresentsANameFactory is good naming.  I of course dislike this terribly.  

We also tend to base these things on feelings of what we like or not.  This is because are brains are hardwired from what we have done before and to conserve energy our brains like to follow this path.  

So I'd like to take these discussion down a more evidence based path of discussion.  Let's take a kinda first principals approach, break down what are we trying to accomplish with our naming.  Then what this naming is doing and see if it meets our goals.  If it doesn't meet them, then let's talk about what would meet them.  

Here is what I see as the goals with our naming:

- Clear with what something does with a name
- Clear across multiple language implementations
- Easy to find the starting point

The first is kind of a two prong problem.  ConfidentialClientApplication() has no real meaning.  It is a confusing term.  app and server would have been better labels.  But that is already done.  So really it is now about: is the name clear if you know the terminology of Go and MSAL.

The second is, if you have used MSAL before, is the name clear enough within tight package contents, to be not have to spend a lot of cognitive energy making the translation.

To accomplish these we use a combination of:
- directory naming
- package naming
- type names and a type system
- small packages with few types (to make searches fast)

With this said, let's pretend I know Go and I've done MSAL from Python.  

My first step is to get my application client.  Let's say its the "confidential one".  Where would that be?

apps/ (no other places to be)

Okay, there is no apps package and I need confidential, is it clear where to find it from this:

apps/
  confidential/
  public/

So I see confidential, now, how do I get the client.  In Go, this is usually done with New().  If I look at the package, is it clear how to do that? 

To answer that, we have 2 of the 3 things I think we need:
- New() (no extra words) is generally used to indicate the main type of a package
- Is the godoc going to be short so there is very little to look through
- Is there a godoc example

I believe we are a two out of three here.  We use New(), we have 2 types and 3 constructors total.  The Godoc will be about a page in length.  

What we fail at is a godoc example.  A good one you copy and paste what is there for your own use.  We don't have one yet, but we will in the future

With this rational, do you believe that the naming is faulty, confusing or unclear to an engineer who has Go experience and MSAL experience?  If so is that lack of clarity for a moment (like 1 second) or is just so confusing that they are lost?  Would our devs have problems figuring it out?

Reply:
I agree that framed within the context of the module as a whole the naming makes sense.  It's just when I look at `confidential.New()` it seems very vague; perhaps that's an unfair comparison though (side note: while I have an appreciation for Go's preference for terse names [Java/C# are just too wordy], I have some bias for more descriptive names based on years of C/C++ programming where I sometimes had to debug in production environments without IDEs and other niceties; doesn't really translate here).

Ultimately, with proper documentation and examples I think customers will use what we provide and move on.

Reply:
ignoring all the goodness of go and conciseness... I prefer the more wordy version as suggested above:
client := msal.NewConfidentialClientApplication()
it's consistent with the naming across libraries.

Reply:
I think it would be helpful if there was a list of type names that _must_ be uniform across languages (maybe not 100% identical, but similar enough that it's obvious).